File: src/test/java/itmo/high_perf_sys/chat/service/MessageServiceTest.java
package itmo.high_perf_sys.chat.service;

import itmo.high_perf_sys.chat.entity.Chat;
import itmo.high_perf_sys.chat.entity.Message;
import itmo.high_perf_sys.chat.repository.chat.MessageRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
public class MessageServiceTest {

    @Mock
    private MessageRepository messageRepository;

    @InjectMocks
    private MessageService messageService;

    private Message message;

    @BeforeEach
    public void setUp() {
        message = new Message();
        message.setId(UUID.randomUUID());
        message.setChatId(new Chat());
        message.setText("Test Message");
        message.setPhoto(new byte[0]);
    }

    @Test
    public void testSave() {
        messageService.save(message);
        verify(messageRepository, times(1)).save(message);
    }

    @Test
    public void testFindLastByChatId() {
        UUID chatId = UUID.randomUUID();
        when(messageRepository.findLastByChatId(chatId)).thenReturn(Optional.of(message));

        Optional<Message> result = messageService.findLastByChatId(chatId);

        assertTrue(result.isPresent());
        assertEquals(message, result.get());
    }

    @Test
    public void testFindLastByChatIdNotFound() {
        UUID chatId = UUID.randomUUID();
        when(messageRepository.findLastByChatId(chatId)).thenReturn(Optional.empty());

        Optional<Message> result = messageService.findLastByChatId(chatId);

        assertFalse(result.isPresent());
    }

    @Test
    public void testFindByTextContainingAndChatId() {
        UUID chatId = UUID.randomUUID();
        Pageable pageable = PageRequest.of(0, 10);
        Page<Message> page = new PageImpl<>(List.of(message));
        when(messageRepository.findByTextContainingAndChatId(chatId, "Test", pageable)).thenReturn(page);

        Page<Message> result = messageService.findByTextContainingAndChatId(chatId, "Test", pageable);

        assertEquals(page, result);
    }

    @Test
    public void testFindByChatId() {
        UUID chatId = UUID.randomUUID();
        Pageable pageable = PageRequest.of(0, 10);
        Page<Message> page = new PageImpl<>(List.of(message));
        when(messageRepository.findByChatId(chatId, pageable)).thenReturn(page);

        Page<Message> result = messageService.findByChatId(chatId, pageable);

        assertEquals(page, result);
    }
}

File: src/test/java/itmo/high_perf_sys/chat/service/MessageServiceTest.kt
package itmo.high_perf_sys.chat.service

import itmo.high_perf_sys.chat.entity.Message
import itmo.high_perf_sys.chat.repository.chat.MessageRepository
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.MockitoAnnotations
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.junit.Assert.*
import java.util.*


class MessageServiceTest {

    @Mock
    private lateinit var messageRepository: MessageRepository

    @InjectMocks
    private lateinit var messageService: MessageService

    @BeforeEach
    fun setUp() {
        MockitoAnnotations.openMocks(this)
    }

    @Test
    fun testSaveMessage() {
        val message = Message().apply {
            id = UUID.randomUUID()
            text = "Test Message"
        }

        messageService.save(message)

        verify(messageRepository, times(1)).save(message)
    }

    @Test
    fun testFindLastByChatId() {
        val chatId = UUID.randomUUID()
        val message = Message().apply {
            id = UUID.randomUUID()
            text = "Last Message"
        }

        `when`(messageRepository.findLastByChatId(chatId)).thenReturn(Optional.of(message))

        val result = messageService.findLastByChatId(chatId)

        assertNotNull(result)
        assertEquals(message, result.get())
    }

    @Test
    fun testFindLastByChatIdNotFound() {
        val chatId = UUID.randomUUID()

        `when`(messageRepository.findLastByChatId(chatId)).thenReturn(Optional.empty())

        val result = messageService.findLastByChatId(chatId)

        assertNotNull(result)
        assertEquals(false, result.isPresent)
    }

    @Test
    fun testFindByTextContainingAndChatId() {
        val chatId = UUID.randomUUID()
        val request = "Test"
        val pageable: Pageable = PageRequest.of(0, 10)
        val messages = listOf(
            Message().apply {
                id = UUID.randomUUID()
                text = "Test Message 1"
            },
            Message().apply {
                id = UUID.randomUUID()
                text = "Test Message 2"
            }
        )

        `when`(messageRepository.findByTextContainingAndChatId(chatId, request, pageable))
            .thenReturn(PageImpl(messages))

        val result = messageService.findByTextContainingAndChatId(chatId, request, pageable)

        assertNotNull(result)
        assertEquals(2, result.content.size)
    }

    @Test
    fun testFindByChatId() {
        val chatId = UUID.randomUUID()
        val pageable: Pageable = PageRequest.of(0, 10)
        val messages = listOf(
            Message().apply {
                id = UUID.randomUUID()
                text = "Message 1"
            },
            Message().apply {
                id = UUID.randomUUID()
                text = "Message 2"
            }
        )

        `when`(messageRepository.findByChatId(chatId, pageable)).thenReturn(PageImpl(messages))

        val result = messageService.findByChatId(chatId, pageable)

        assertNotNull(result)
        assertEquals(2, result.content.size)
    }
}
